File: 141_Linked_List_cycle.py
#two pointer approach, one fast pointer and one slow pointer
class Solution:
    def hasCycle(self, head: ListNode) -> bool:
        # Floyd Cycle Detection Algorithm, Tortoise and Hare Algorithm, Fast and slow pointers
        walker,runner = head,head
        #runner = head
        while runner != None and runner.next != None:
            walker = walker.next
            runner = runner.next.next
            if walker == runner:
                return True
        return False

File: 141_Linked_List_cycle.cpp
// two pointer approach one fast pointer and one slow pointer
class Solution {
public:
   bool hasCycle(ListNode *head) {
	if (!head)
		return false;
	ListNode* slow = head;
	ListNode* fast = head->next;
	while (fast && fast->next) {
		if (slow == fast)
			return true;
		slow = slow->next;
		fast = fast->next->next;
	}
	return false;
}
};